File: 0451-sort-characters-by-frequency/0451-sort-characters-by-frequency.py
class Solution:
    def frequencySort(self, s: str) -> str:
        cache = {}
        for character in s:
            if character not in cache:
                cache[character] = 0
            cache[character] += 1
        
        sortedArray = sorted(cache.items(), key=lambda x: x[1], reverse=True)
        result = "".join([character * freq for character, freq in sortedArray])
        return result

File: 0451-sort-characters-by-frequency/0451-sort-characters-by-frequency.cpp
class Solution {
public:
    string frequencySort(string s) {
        unordered_map<char,int> freq;
        for(char c:s) {
            freq[c]++;
        }
        
        vector<string> bucket(s.size()+1, "");
        for(auto& it:freq) {
            int n = it.second;
            char c = it.first;
            bucket[n].append(n, c);
        }
        
        string res;
        for(int i=s.size(); i>0; i--) {
            if(bucket[i].empty()) {
                continue;
            }
            res.append(bucket[i]);
        }
        return res;
    }
};